crates/pink/runtime/src/runtime.rs

```diff
@@ -161,3 +162,17 @@ fn detect_parameter_changes() {
         <PinkRuntime as Config>::MaxStorageKeyLen::get(),
     ));
 }
+
+/// Call on_genesis for all pallets. This would put the each pallet's storage version into
+/// the frame_support::StorageVersion.
+pub fn on_genesis() {
+    <AllPalletsWithSystem as frame_support::traits::OnGenesis>::on_genesis();
+}
+
+/// Call on_runtime_upgrade for all pallets
```

Comment:
Note that although every pallet has an `on_runtime_upgrade` hook for migration, on the upstream, it's preferred to specifically define the runtime level migrations in this way: `type Migrations = (A, B, C)`

The reason is that, pallet defined migrations is implicit. The runtime maintainer may not notice the migration. However, usually migrations are highly depends on the runtime (e.g. Phala vs Khala). So in the end, Parity simply prefer to run migrations only defined in the runtime and it's checked in each release. They have nearly abandoned the pallet-based migration.

Reply:
Here's the example:

https://github.com/paritytech/polkadot/blob/851071470841fb49d5ab3fe7d4f43c6cef5e489f/runtime/kusama/src/lib.rs#L1506-L1507

Reply:
Yes, here we actually defined our custom migration: `type Migrations = (Migration<PinkRuntime>, AllPalletsWithSystem);` where `Migration` is `pallet_contracts::Migration`.
So the `Migrations:: on_runtime_upgrade` will first call the explicit migration from `pallet_contracts` and then the hook-based migrations.
The order is following the default substrate node implementation. We have a few pallets in our tiny runtime and only the `pallet_contracts` needs an explicit migration.